scopesim/server/database.py

```diff
@@ -1,48 +1,287 @@
+# -*- coding: utf-8 -*-
 """
 Functions to download instrument packages and example data
 """
-import json
 import re
-import shutil
-import os
-import urllib.request
-import zipfile
+from zipfile import ZipFile
 import logging
-from urllib3.exceptions import HTTPError
+from datetime import date
+from warnings import warn
+from pathlib import Path
+from typing import Optional, Union, List, Tuple, Set, Dict
+# Python 3.8 doesn't yet know these things.......
+# from collections.abc import Iterator, Iterable, Mapping
+from typing import Iterator, Iterable, Mapping
+from shutil import get_terminal_size
+
+from urllib.error import HTTPError
+from urllib3.exceptions import HTTPError as HTTPError3
+from more_itertools import first, last, groupby_transform
 
-import yaml
 import requests
+from requests.packages.urllib3.util.retry import Retry
+from requests.adapters import HTTPAdapter
+from requests_cache import CachedSession
 import bs4
-from astropy.utils.data import download_file
+from tqdm import tqdm
+# from tqdm.contrib.logging import logging_redirect_tqdm
+# put with logging_redirect_tqdm(loggers=all_loggers): around tqdm
 
 from scopesim import rc
+from .github_utils import download_github_folder
+from .example_data_utils import (download_example_data, list_example_data,
+                                 get_server_elements)
 
+GrpVerType = Mapping[str, Iterable[str]]
+GrpItrType = Iterator[Tuple[str, List[str]]]
+
+
+width, _ = get_terminal_size((50, 20))
+width = int(.8 * width)
+bar_width = max(width - 50, 10)
+tqdm_kwargs = {
+    "bar_format": f"{{l_bar}}{{bar:{bar_width}}}{{r_bar}}{{bar:-{bar_width}b}}",
+    "colour": "green"
+    }
```

Comment:
Perhaps put these in a function?

* I frequently resize my terminal, which would now screw up the progress bars right?
* If someone where to do `from scopesim.server.database import *`, they would now get `width` imported.
* Now `tqdm_kwargs` is a global. If we would later parallelize the downloads, would that mess up the progress bars?

Reply:
Indeed, you're right! I took these from a script in teutoburg/prepipy, where it doesn't really matter, because it's executed as a script, not imported as a module. But I totally agree that it's a different story here. Isolating this in a simple aux function returning `tqdm_kwargs`, which is called every time we need a progress bar, should solve this.